libraries/AC_PrecLand/AC_PrecLand_StateMachine.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include <AP_Math/AP_Math.h>
+#include <AC_PrecLand/AC_PrecLand.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_InertialNav/AP_InertialNav.h>
+
+// This class constantly monitors what the status of the landing target is
+// If it is not in sight, depending on user parameters, it decides what measures can be taken to bring the target back in sight
+// If the target has been lost recently, the vehicle might try to retry the landing by going to the last known location of the target/going to the location where the target was last detected
+// If we have no clue where the target might be, then failsafe measures are activated
+// Failsafe measures can include stopping completely (hover), or landing vertically
+class AC_PrecLand_StateMachine {
+public:
+
+    // Constructor
+    AC_PrecLand_StateMachine(const AP_InertialNav& inav);
+
+    // Do not allow copies
+    AC_PrecLand_StateMachine(const AC_PrecLand_StateMachine &other) = delete;
+    AC_PrecLand_StateMachine &operator=(const AC_PrecLand_StateMachine&) = delete;
+
+    // Initialize the state machine. This is called everytime vehicle switches mode
+    void init();
+
+    // Current status of the precland state machine
+    enum class Status: uint8_t {
+        ERROR = 0,               // Unkown error
+        DESCEND,                 // No action is required, just descend vertically
+        RETRYING,                // Vehicle is attempting to retry landing
+        FAILSAFE                 // Switch to prec landing failsafe
+    };
+
+    // FailSafe action needed
+    enum class FailSafeAction: uint8_t {
+        HOLD_POS = 0,            // Hold the current position of the vehicle
+        DESCEND                  // Descend vertically
+    };
+
+    // Run Prec Land State Machine. During Prec Landing, we might encounter four scenarios:
+    // 1. We had the target in sight, but have lost it now. 2. We never had the target in sight and user wants to land.
+    // 3. We have the target in sight and can continue landing. 4. The sensor is out of range
+    // This method deals with all of these scenarios
+    // Returns the action needed to be done by the vehicle.
+    // Parameters: Vector3f "retry_pos_cm" is filled with the required location if we need to retry landing.
+    Status update_precland_state_machine(Vector3f &retry_pos_cm);
+
+    // Decide what the next action is going to be if we have hit a failsafe
+    // At the moment this method only allows you to stop in air permanently, or land vertically
+    // The FS action is returned as a enum value
+    // Failsafe will only trigger as a last resort
+    FailSafeAction do_failsafe_actions();
```

Comment:
I think "do_failsafe_actions()" might be better off called "get_failsafe_action".
In the comments we should make it clear that this should only be called if update_precland_state_machine() returns FAILSAFE.  I just want the users of this library to understand that this method doesn't actually "do" anything.. instead it tells the vehicle code what it should do.